File: golang/number_islands.go
func numIslands(grid [][]byte) int {
	islands := 0
	seen := map[string]bool{}

	for row := 0; row < len(grid); row++ {
		for col := 0; col < len(grid[0]); col++ {
			// skip if 0
			if grid[row][col] == '0' {
				continue
			}

			// seen already => skip
			if seen[fmt.Sprint(row, col)] {
				continue
			} else {
				// other we found a new island
				islands++
			}

			// perform dfs and mark all connected land
			helper(grid, seen, row, col)
		}
	}

	return islands
}

func helper(grid [][]byte, seen map[string]bool, row, col int) {
	// base case
	if row < 0 || row >= len(grid) || col < 0 || col >= len(grid[0]) {
		return
	}
	if grid[row][col] == '0' {
		return
	}
	if seen[fmt.Sprint(row, col)] {
		return
	}

	// mark visited
	seen[fmt.Sprint(row, col)] = true

	// move along every direction to search for connected land
	helper(grid, seen, row-1, col)
	helper(grid, seen, row+1, col)
	helper(grid, seen, row, col-1)
	helper(grid, seen, row, col+1)
}

// instead of a seen map, you could have directly marked the cell with 'x' or something similar

File: typescript/number_islands.ts
function numIslands(grid: string[][]): number {
	// keep island count and perform dfs to mark all 1's that are adjacent
	let islands = 0

	for (let row = 0; row < grid.length; row++) {
		for (let col = 0; col < grid[0].length; col++) {
			let cell = grid[row][col]

			// skip if not island start
			if (cell == '0') continue

			// perform dfs to mark entire island and increment count
			if (cell == '1') {
				helper(grid, row, col)
				islands++
			}
		}
	}

	return islands
}

function helper(grid: string[][], row: number, col: number) {
	// base case
	if (row < 0 || row >= grid.length || col < 0 || col >= grid[0].length || grid[row][col] == '0') return

	// mark visited
	grid[row][col] = '0'

	// traverse in every cardinal direction
	helper(grid, row - 1, col)
	helper(grid, row + 1, col)
	helper(grid, row, col - 1)
	helper(grid, row, col + 1)
}
